src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1050,67 +1050,69 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
-                localSymbol.SetSpecificallyNotPinned();
-
-                // UNDONE: ExpressionBinder::CheckFieldUse (something about MarshalByRef)
-            }
-            else if (initializerType.IsArray())
+            else
             {
-                // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
-
-                var arrayType = (ArrayTypeSymbol)initializerType;
-                TypeSymbol elementType = arrayType.ElementType;
-
+                TypeSymbol elementType;
                 bool hasErrors = false;
-                if (elementType.IsManagedType)
+
+                if (initializerType.SpecialType == SpecialType.System_String)
                 {
-                    Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
-                    hasErrors = true;
+                    elementType = this.GetSpecialType(SpecialType.System_Char, diagnostics, initializerSyntax);
+                    Debug.Assert(!elementType.IsManagedType);
                 }
+                else if (initializerType.IsArray())
+                {
+                    // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
+                    var arrayType = (ArrayTypeSymbol)initializerType;
+                    elementType = arrayType.ElementType;
 
-                initializerOpt = GetFixedLocalCollectionInitializer(initializerOpt, elementType, declType, hasErrors, diagnostics);
-            }
-            else
-            {
-                if (!initializerOpt.HasAnyErrors)
+                    if (elementType.IsManagedType)
+                    {
+                        Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
+                        hasErrors = true;
+                    }
+                }
+                else if (!initializerOpt.HasAnyErrors)
                 {
+                    elementType = initializerOpt.Type;
                     switch (initializerOpt.Kind)
                     {
                         case BoundKind.AddressOfOperator:
-                            // OK
-                            break;
-                        case BoundKind.Conversion:
-                            // The following assertion would not be correct because there might be an implicit conversion after (above) the explicit one.
-                            //Debug.Assert(((BoundConversion)initializerOpt).ExplicitCastInCode, "The assignment conversion hasn't been applied yet, so this must be from source.");
-
-                            // NOTE: Dev10 specifically doesn't report this error for the array or string cases.
-                            Error(diagnostics, ErrorCode.ERR_BadCastInFixed, initializerSyntax);
+                            elementType = ((BoundAddressOfOperator)initializerOpt).Operand.Type;
```

Comment:
I have checked that any path that create BoundAddressOf will not allow the Operand to have null type. There are not that many places when that is created actually. They all force at least an error type, since the node generally has a pointer type of that.

We already have some tests with typeless operands used in `fixed`
I will add some more scenarios to be sure. 

---
In reply to: [125144198](https://github.com/dotnet/roslyn/pull/20548#discussion_r125144198) [](ancestors = 125144198,125144136,125143535)